File: server/src/modules/system/role/role.service.ts
import { BaseService } from "@/core/base.service";
import { Result } from "@/core/result";
import prisma from "@/core/prisma";

class RoleService extends BaseService {
  constructor() {
    super("role");
  }

  async pageList(page: number, pageSize: number, params: any) {
    const where: any = {};
    if (params.roleName) where.roleName = { contains: params.roleName };
    if (params.roleCode) where.roleCode = { contains: params.roleCode };
    if (params.description) where.description = { contains: params.description };
    if (params.enabled !== undefined) where.enabled = params.enabled;

    const skip = (page - 1) * pageSize;
    const [list, total] = await Promise.all([
      this.model.findMany({
        skip,
        take: pageSize,
        where,
        orderBy: { createTime: "desc" },
      }),
      this.model.count({ where }),
    ]);
    return { list, total };
  }

  // 给角色分配菜单权限
  async updateMenus(roleId: number, menuIds: number[]) {
    await this.model.update({
      where: { id: roleId },
      data: {
        menus: { set: menuIds.map((mid) => ({ id: mid })) },
      },
    });
    return Result.success(null, "权限分配成功");
  }

  // 获取某个角色拥有的菜单 ID 列表（回显用）
  async getMenuIdsByRoleId(roleId: number) {
    const role = await this.model.findUnique({
      where: { id: roleId },
      include: { menus: true },
    });
    const menuIds = role?.menus.map((m: any) => m.id) || [];
    return Result.success(menuIds);
  }

  // 分配数据权限
  async updateDataScope(roleId: number, dataScope: number, deptIds: number[] = []) {
    const deptsUpdate =
      dataScope === 2 ? { set: deptIds.map((id) => ({ id })) } : { set: [] };
    await this.model.update({
      where: { id: roleId },
      data: { dataScope, depts: deptsUpdate },
    });
    return Result.success(null, "数据权限分配成功");
  }
}

export default new RoleService();

File: java-server/src/main/java/com/todothink/modules/system/role/service/RoleService.java
package com.todothink.modules.system.role.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.LambdaUpdateWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.todothink.core.exception.BusinessException;
import com.todothink.modules.system.entity.Role;
import com.todothink.modules.system.mapper.RelationMapper;
import com.todothink.modules.system.mapper.RoleMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class RoleService {

    private final RoleMapper roleMapper;
    private final RelationMapper relationMapper;

    public RoleService(RoleMapper roleMapper, RelationMapper relationMapper) {
        this.roleMapper = roleMapper;
        this.relationMapper = relationMapper;
    }

    public Map<String, Object> pageList(int page, int pageSize, Map<String, Object> params) {
        LambdaQueryWrapper<Role> wrapper = new LambdaQueryWrapper<Role>();
        if (StringUtils.hasText((String) params.get("roleName"))) {
            wrapper.like(Role::getRoleName, params.get("roleName"));
        }
        if (StringUtils.hasText((String) params.get("roleCode"))) {
            wrapper.like(Role::getRoleCode, params.get("roleCode"));
        }
        if (StringUtils.hasText((String) params.get("description"))) {
            wrapper.like(Role::getDescription, params.get("description"));
        }
        if (params.get("enabled") != null) {
            wrapper.eq(Role::getEnabled, params.get("enabled"));
        }
        wrapper.orderByDesc(Role::getCreateTime);

        Page<Role> pageResult = roleMapper.selectPage(new Page<Role>(page, pageSize), wrapper);
        Map<String, Object> result = new HashMap<String, Object>();
        result.put("list", pageResult.getRecords());
        result.put("total", pageResult.getTotal());
        return result;
    }

    @Transactional
    public Role add(Role role) {
        if (Boolean.TRUE.equals(role.getIsDefaultRole())) {
            clearDefaultRole(null);
        }
        roleMapper.insert(role);
        return role;
    }

    @Transactional
    public Role update(Integer id, Role role) {
        if (Boolean.TRUE.equals(role.getIsDefaultRole())) {
            clearDefaultRole(id);
        }
        role.setId(id);
        roleMapper.updateById(role);
        return roleMapper.selectById(id);
    }

    public void delete(Integer id) {
        roleMapper.deleteById(id);
    }

    @Transactional
    public void updateMenus(Integer roleId, List<Integer> menuIds) {
        relationMapper.deleteRoleMenus(roleId);
        if (menuIds != null && !menuIds.isEmpty()) {
            relationMapper.insertRoleMenus(roleId, menuIds);
        }
    }

    public List<Integer> getMenuIdsByRoleId(Integer roleId) {
        return relationMapper.selectMenuIdsByRoleId(roleId);
    }

    @Transactional
    public void updateDataScope(Integer roleId, Integer dataScope, List<Integer> deptIds) {
        Role update = new Role();
        update.setId(roleId);
        update.setDataScope(dataScope);
        roleMapper.updateById(update);
        relationMapper.deleteRoleDepts(roleId);
        if (dataScope != null && dataScope == 2 && deptIds != null && !deptIds.isEmpty()) {
            relationMapper.insertRoleDepts(roleId, deptIds);
        }
    }

    private void clearDefaultRole(Integer excludeId) {
        LambdaUpdateWrapper<Role> wrapper = new LambdaUpdateWrapper<Role>()
                .eq(Role::getIsDefaultRole, true)
                .isNull(Role::getDeletedAt)
                .set(Role::getIsDefaultRole, false);
        if (excludeId != null) {
            wrapper.ne(Role::getId, excludeId);
        }
        roleMapper.update(null, wrapper);
    }
}
